### PR TITLE
Rewrite spec.md for 1.0.0

### PR DESCRIPTION
**Pull request by squeed**
The protocol is unchanged, but the specification was badly in need of a rewrite.

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
-Runtimes should then look for this executable in a list of predefined directories (the list of directories is not prescribed by this specification). Once found, it must invoke the executable using the following environment variables for argument passing:
+A network configuration consists of a JSON object with the following keys:
 
-- `CNI_COMMAND`: indicates the desired operation; `ADD`, `DEL`, `CHECK`, or `VERSION`.
-- `CNI_CONTAINERID`: Container ID
-- `CNI_NETNS`: Path to network namespace file
-- `CNI_IFNAME`: Interface name to set up; if the plugin is unable to use this interface name it must return an error
-- `CNI_ARGS`: Extra arguments passed in by the user at invocation time. Alphanumeric key-value pairs separated by semicolons; for example, "FOO=BAR;ABC=123"
-- `CNI_PATH`: List of paths to search for CNI plugin executables. Paths are separated by an OS-specific list separator; for example ':' on Linux and ';' on Windows
+- `cniVersion` (string): [Semantic Version 2.0](https://semver.org) of CNI specification to which this configuration list and all the individual configurations conform. Currently "1.0.0"
```
Comment (bboreham) [inline SPEC.md:66]: I would say "semver" is more commonly understood than "Semantic Version".

File: SPEC.md [hunk 2]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
+- `name` (string): Network name. This should be unique across all containers on the host (or other administrative domain).  Must start with a alphanumeric character, optionally followed by any combination of one or more alphanumeric characters, underscore , dot (.) or hyphen (-).
```
Comment (bboreham) [inline SPEC.md:67]: s/a alphanumeric/an alphanumeric/

File: SPEC.md [hunk 3]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
 
-For the purposes of this proposal, we define two terms very specifically:
-- _container_ can be considered synonymous with a [Linux _network namespace_][namespaces]. What unit this corresponds to depends on a particular container runtime implementation: for example, in implementations of the [App Container Spec][appc-github] like rkt, each _pod_ runs in a unique network namespace. In [Docker][docker], on the other hand, network namespaces generally exist for each separate Docker container.
-- _network_ refers to a group of entities that are uniquely addressable that can communicate amongst each other. This could be either an individual container (as specified above), a machine, or some other network device (e.g. a router). Containers can be conceptually _added to_ or _removed from_ one or more networks.
+For the purposes of this proposal, we define three terms very specifically:
+- _container_ can be considered synonymous with a [Linux _network namespace_][namespaces], though the actual isolation technology is not defined by the specification. What unit this corresponds to depends on a particular container runtime implementation.
```
Comment (bboreham) [inline SPEC.md:30]: I think we should be explicit that non-containers such as VMs are allowed, and the fields are just re-purposed

File: SPEC.md [hunk 4]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
+- `CNI_COMMAND`
+
+
+## Section 3: Execution of Network Configurations
+
+This section describes how a container runtime interprets a network configuration (as defined in section 1) and executes plugins accordingly. A runtime may wish to _add_, _delete_, or _check_ a network configuration in a container. This results in a series of plugin `ADD`, `DELETE`, or `CHECK` executions, correspondingly. This section also defines how a network configuration is transformed and provided to the plugin.
+
+The operation of a network configuration on a container is called an _attachment_. An attachment may be uniquely identified by the `(CNI_CONTAINERID, CNI_IFNAME)` tuple.
+
+Runtimes
```
Comment (bboreham) [inline SPEC.md:291]: What is this line doing?

File: SPEC.md [hunk 5]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
+- **Namespace**: A reference to the container's "isolation domain". If using network namespaces, then a path to the network namespace (e.g. `/run/netns/[nsname]`). During execution, always set as the `CNI_NETNS` parameter.
+- **Container interface name**: Name of the interface to create inside the container. During execution, always set as the `CNI_IFNAME` parameter.
+- **Generic Arguments**: Extra arguments, in the form of key-value string pairs, that are relevant to a specific attachment.  During execution, always set as the `CNI_ARGS` parameter as well as provided in the generated configuration, see below.
+- **Capability Arguments**: These are also key-value pairs. The key is a string, whereas the value is any JSON-serializable type. The keys and values are defined by [convention](CONVENTIONS.md).
+
+Furthermore, the runtime must be provided a list of paths to search for CNI plugins. This must also be provided to plugins during execution via the `CNI_PATH` environment variable.
+
+### Adding an attachment
+To attach create a network attachment, the runtime must first create the isolation domain.
```
Comment (bboreham) [inline SPEC.md:315]: s/attach create/attach/

**Comment by squeed** [comment]
CNI maintainers: any more changes?

File: SPEC.md [hunk 6]
```diff
@@ -51,401 +25,700 @@ Released versions of the spec are available as Git tags.
...
+A network configuration consists of a JSON object with the following keys:
 
-- `CNI_COMMAND`: indicates the desired operation; `ADD`, `DEL`, `CHECK`, or `VERSION`.
-- `CNI_CONTAINERID`: Container ID
-- `CNI_NETNS`: Path to network namespace file
-- `CNI_IFNAME`: Interface name to set up; if the plugin is unable to use this interface name it must return an error
-- `CNI_ARGS`: Extra arguments passed in by the user at invocation time. Alphanumeric key-value pairs separated by semicolons; for example, "FOO=BAR;ABC=123"
-- `CNI_PATH`: List of paths to search for CNI plugin executables. Paths are separated by an OS-specific list separator; for example ':' on Linux and ';' on Windows
+- `cniVersion` (string): [Semantic Version 2.0](https://semver.org) of CNI specification to which this configuration list and all the individual configurations conform. Currently "1.0.0"
+- `name` (string): Network name. This should be unique across all containers on the host (or other administrative domain).  Must start with an alphanumeric character, optionally followed by any combination of one or more alphanumeric characters, underscore , dot (.) or hyphen (-).
```
Comment (dcbw) [inline SPEC.md:67]: @squeed "across all containers"?

**Review by bboreham — APPROVED** [review]
Excellent work!  Let's merge this and clean up anything else after.

**Review by mars1024 — APPROVED** [review]
/lgtm, a huge and professional job!

File: SPEC.md [hunk 7]
```diff
@@ -51,401 +25,697 @@ Released versions of the spec are available as Git tags.
...
+- `prevResult`: A JSON object, consisting of the result type returned by the "previous" plugin. The meaning of "previous" is defined by the specific operation (_add_, _delete_, or _check_).
+
+The following fields must be **removed** by the runtime:
+- `capabilities`
+
+All other fields should be passed through unaltered.
+
+#### Deriving `runtimeConfig`
+
+Whereas `args` and CNI_ARGS are provided to all plugins, with no indication if they are going to be consumed, _Capability arguments_ need to be declared explicitly in configuration. The runtime, thus, can determine if a given network configuration supports a specific _capability_. Capabilities are not defined by the specification - rather, they are documented [conventions](CONVENTIONS.md).
```
Comment (s1061123) [inline SPEC.md:372]: If `args` and `CNI_ARGS` gives different parameters on same argument, what is happen? Is that in scope of SPEC?

```
CNI_ARGS=argA=bar

{
    "cniVersion": "1.0.0",
    "name": "dbnet",
    "type": "bridge",
    "bridge": "cni0",
    "keyA": ["some more", "plugin specific", "configuration"],
    "args": {"argA": "foo"},
    "ipam": {
        "type": "host-local",
        "subnet": "10.1.0.0/16",
        "gateway": "10.1.0.1"
    },
    "dns": {
        "nameservers": [ "10.1.0.1" ]
    }
}
```

**Comment by passcod** [comment]
Shouldn't the table of contents be re-added?

**Comment by squeed** [comment]
Good point, added a TOC.

**Pull request merged**